File: Scala/src/main/scala/TextbookTrack/Chapter04/BA4M.scala
package TextbookTrack.Chapter04

object BA4M {
  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def calcOccurrences[T](xs: List[T]): Map[T, Int] =
    xs.groupMapReduce(identity)(_ => 1)(_ + _)

  private def readPositiveDifferences(line: String): Map[Int, Int] = {
    val differences: List[Int] = convertToIntList(line)
    calcOccurrences(differences.filter(_ > 0))
  }

  private def calcPositiveDifferences(xs: List[Int], elem: Int): Map[Int, Int] =
    calcOccurrences(xs.map(x => math.abs(elem - x)))

  private def isSubset(delta: Map[Int, Int], diffs: Map[Int, Int]): Boolean =
    delta.forall{ case (item, count) => diffs.getOrElse(item, 0) >= count }

  private def remove(differences: Map[Int, Int], items: Map[Int, Int]): Map[Int, Int] =
    items.foldLeft(differences){
      case (acc, (item, count)) =>
        val currentCount: Int = acc.getOrElse(item, 0)
        if (currentCount == count) acc.removed(item)
        else acc.updated(item, currentCount - count)
    }

  private def insertLargestValue(xs: List[Int], differences: Map[Int, Int], y: Int, m: Int): Option[List[Int]] = {
    val delta: Map[Int, Int] = calcPositiveDifferences(xs, y)
    if (isSubset(delta, differences)) {
      val updatedDifferences: Map[Int, Int] = remove(differences, delta)
      solve(y :: xs, updatedDifferences, m)
    } else None
  }

  private def solve(xs: List[Int], differences: Map[Int, Int], m: Int): Option[List[Int]] = {
    if (differences.isEmpty) Some(xs)
    else {
      val y: Int = differences.keysIterator.max
      lazy val result1: Option[List[Int]] = insertLargestValue(xs, differences, y, m)
      lazy val result2: Option[List[Int]] = insertLargestValue(xs, differences, m - y, m)
      if (result1.nonEmpty) result1
      else if (result2.nonEmpty) result2
      else None
    }
  }

  def solveTurnpikeProblem(differences: Map[Int, Int]): List[Int] = {
    val m: Int = differences.keysIterator.max
    val updatedDifferences: Map[Int, Int] = remove(differences, Map(m -> 1))
    val distances: List[Int] = List(m, 0)
    solve(distances, updatedDifferences, m) match {
      case Some(result) => result.sorted
      case None => throw new Exception("No solution has been found.")
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val differences: Map[Int, Int] = readPositiveDifferences(reader.next())
    val result: List[Int] = solveTurnpikeProblem(differences)
    println(result.mkString(" "))
  }
}

File: Python/textbook_track/chapter04/ba4m.py
# Solve the Turnpike Problem
import sys

from collections import Counter


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_positive_differences(line):
    diffs = convert_to_intlist(line)
    return Counter(filter(lambda x: x > 0, diffs))


def calc_positive_differences(xs, elem):
    return Counter([abs(elem - x) for x in xs])


def is_subset(delta, diffs):
    return all(diffs.get(item, 0) >= count for item, count in delta.items())


def remove(differences, items):
    for item, count in items.items():
        current_count = differences.get(item, 0)
        if count == current_count:
            _ = differences.pop(item, None)
        else:
            differences.subtract({item: count})


def insert_largest_value(xs, differences, y, m):
    delta = calc_positive_differences(xs, y)
    if is_subset(delta, differences):
        xs.append(y)
        remove(differences, delta)
        result = solve(xs, differences, m)
        if result is not None:
            return result
        _ = xs.pop(-1)
        differences.update(delta)
    return None


def solve(xs, differences, m):
    if not differences:
        return tuple(xs)

    y = max(differences)
    result = insert_largest_value(xs, differences, y, m)
    if result is not None:
        return result

    result = insert_largest_value(xs, differences, m - y, m)
    if result is not None:
        return result

    return None


def solve_turnpike_problem(differences):
    m = max(differences)
    remove(differences, {m: 1})
    distances = [0, m]
    result = solve(distances, differences, m)
    return sorted(result)


def main():
    data = sys.stdin.read().splitlines()
    differences = read_positive_differences(data[0])
    result = solve_turnpike_problem(differences)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()
